File: app/src/main/java/com/micewine/emu/adapters/AdapterGame.java
package com.micewine.emu.adapters;

import static com.micewine.emu.activities.GeneralSettingsActivity.SELECTED_BOX64;
import static com.micewine.emu.activities.GeneralSettingsActivity.SELECTED_VULKAN_DRIVER;
import static com.micewine.emu.activities.MainActivity.ACTION_RUN_WINE;
import static com.micewine.emu.activities.MainActivity.preferences;
import static com.micewine.emu.core.RatPackageManager.listRatPackagesId;
import static com.micewine.emu.fragments.DebugSettingsFragment.availableCPUs;
import static com.micewine.emu.fragments.ShortcutsFragment.MESA_DRIVER;
import static com.micewine.emu.fragments.ShortcutsFragment.getBox64Preset;
import static com.micewine.emu.fragments.ShortcutsFragment.getBox64Version;
import static com.micewine.emu.fragments.ShortcutsFragment.getCpuAffinity;
import static com.micewine.emu.fragments.ShortcutsFragment.getD3DXRenderer;
import static com.micewine.emu.fragments.ShortcutsFragment.getDXVKVersion;
import static com.micewine.emu.fragments.ShortcutsFragment.getDisplaySettings;
import static com.micewine.emu.fragments.ShortcutsFragment.getEnableDInput;
import static com.micewine.emu.fragments.ShortcutsFragment.getEnableXInput;
import static com.micewine.emu.fragments.ShortcutsFragment.getSelectedVirtualControllerPreset;
import static com.micewine.emu.fragments.ShortcutsFragment.getVKD3DVersion;
import static com.micewine.emu.fragments.ShortcutsFragment.getVulkanDriver;
import static com.micewine.emu.fragments.ShortcutsFragment.getVulkanDriverType;
import static com.micewine.emu.fragments.ShortcutsFragment.getWineD3DVersion;
import static com.micewine.emu.fragments.ShortcutsFragment.getWineESync;
import static com.micewine.emu.fragments.ShortcutsFragment.getWineServices;
import static com.micewine.emu.fragments.ShortcutsFragment.getWineVirtualDesktop;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.ColorMatrix;
import android.graphics.ColorMatrixColorFilter;
import android.graphics.Paint;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.FragmentActivity;
import androidx.recyclerview.widget.RecyclerView;

import com.micewine.emu.R;
import com.micewine.emu.activities.EmulationActivity;
import com.micewine.emu.fragments.EditGamePreferencesFragment;

import java.io.File;
import java.util.ArrayList;
import java.util.Arrays;

public class AdapterGame extends RecyclerView.Adapter<AdapterGame.ViewHolder> {
    private final FragmentActivity activity;
    private final float size;
    private final ArrayList<GameItem> gameList;
    private final ArrayList<GameItem> filteredList = new ArrayList<>();

    public AdapterGame(ArrayList<GameItem> gameList, float size, FragmentActivity activity) {
        this.activity = activity;
        this.size = size;
        this.gameList = gameList;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View itemView = LayoutInflater.from(parent.getContext()).inflate(R.layout.adapter_game_item, parent, false);
        itemView.getLayoutParams().width = Math.round(itemView.getLayoutParams().width * size);
        itemView.getLayoutParams().height = Math.round(itemView.getLayoutParams().height * size);
        return new ViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        GameItem item = filteredList.get(position);

        holder.titleGame.setText(item.name);

        File imgFile = new File(item.iconPath);
        if (imgFile.exists() && imgFile.length() > 0) {
            Bitmap imgBitmap = BitmapFactory.decodeFile(item.iconPath);
            if (imgBitmap != null) {
                Bitmap newBitmap;
                if (new File(item.exePath).exists()) {
                    newBitmap = Bitmap.createScaledBitmap(imgBitmap, holder.itemView.getLayoutParams().width - 10, holder.itemView.getLayoutParams().width - 10, false);
                } else {
                    newBitmap = toGrayScale(
                            Bitmap.createScaledBitmap(imgBitmap, holder.itemView.getLayoutParams().width - 10, holder.itemView.getLayoutParams().width - 10, false)
                    );
                }
                holder.gameImage.setImageBitmap(newBitmap);
            }
        } else if (item.iconPath.isEmpty()) {
            holder.gameImage.setImageBitmap(
                    BitmapFactory.decodeResource(activity.getResources(), R.drawable.default_icon)
            );
        } else {
            holder.gameImage.setImageResource(R.drawable.unknown_exe);
        }
    }

    @Override
    public int getItemCount() {
        return filteredList.size();
    }

    private Bitmap toGrayScale(Bitmap bitmap) {
        Bitmap grayScaleBitmap = Bitmap.createBitmap(bitmap.getWidth(), bitmap.getHeight(), Bitmap.Config.ARGB_8888);

        Paint paint = new Paint();
        ColorMatrix colorMatrix = new ColorMatrix();
        colorMatrix.setSaturation(0F);

        ColorMatrixColorFilter filter = new ColorMatrixColorFilter(colorMatrix);
        paint.setColorFilter(filter);

        Canvas canvas = new Canvas(grayScaleBitmap);
        canvas.drawBitmap(bitmap, 0F, 0F, paint);

        return grayScaleBitmap;
    }

    @SuppressLint("NotifyDataSetChanged")
    public void filterList(String name) {
        filteredList.clear();

        for (GameItem item : gameList) {
            if (item.name != null && item.name.toLowerCase().contains(name.toLowerCase())) {
                filteredList.add(item);
            }
        }

        notifyDataSetChanged();
    }

    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {
        public TextView titleGame = itemView.findViewById(R.id.title_game_model);
        public ImageView gameImage = itemView.findViewById(R.id.img_game);

        public ViewHolder(@NonNull View itemView) {
            super(itemView);
            itemView.setOnClickListener(this);
            itemView.setOnLongClickListener(this);
        }

        @Override
        public void onClick(View view) {
            GameItem gameItem = gameList.get(getAdapterPosition());

            selectedGameName = gameItem.name;

            File exeFile =  new File(gameItem.exePath);

            String exePath = gameItem.exePath;
            String exeArguments = gameItem.exeArguments;

            if (!exeFile.exists()) {
                if (gameItem.exePath.equals(activity.getString(R.string.desktop_mode_init))) {
                    exePath = "";
                    exeArguments = "";
                } else {
                    activity.runOnUiThread(() -> Toast.makeText(activity, R.string.executable_file_not_found, Toast.LENGTH_SHORT).show());
                    new EditGamePreferencesFragment(EditGamePreferencesFragment.EDIT_GAME_PREFERENCES, null).show(activity.getSupportFragmentManager(), "");
                    return;
                }
            }

            String driverName = getVulkanDriver(selectedGameName);
            if (driverName.equals("Global")) {
                driverName = (preferences != null ? preferences.getString(SELECTED_VULKAN_DRIVER, "") : "");
            }

            String box64Version = getBox64Version(selectedGameName);
            if (box64Version.equals("Global")) {
                box64Version = (preferences != null ? preferences.getString(SELECTED_BOX64, "") : "");
            }

            int driverType = getVulkanDriverType(selectedGameName);

            Intent runActivityIntent = new Intent(activity, EmulationActivity.class);
            Intent runWineIntent = new Intent(ACTION_RUN_WINE);

            runWineIntent.putExtra("exePath", exePath);
            runWineIntent.putExtra("exeArguments", exeArguments);
            runWineIntent.putExtra("driverName", driverName);
            runWineIntent.putExtra("driverType", driverType);
            runWineIntent.putExtra("box64Version", box64Version);
            runWineIntent.putExtra("box64Preset", getBox64Preset(selectedGameName));
            runWineIntent.putExtra("displayResolution", getDisplaySettings(selectedGameName).get(1));
            runWineIntent.putExtra("virtualControllerPreset", getSelectedVirtualControllerPreset(selectedGameName));
            runWineIntent.putExtra("d3dxRenderer", getD3DXRenderer(selectedGameName));
            runWineIntent.putExtra("wineD3D", getWineD3DVersion(selectedGameName));
            runWineIntent.putExtra("dxvk", getDXVKVersion(selectedGameName));
            runWineIntent.putExtra("vkd3d", getVKD3DVersion(selectedGameName));
            runWineIntent.putExtra("esync", getWineESync(selectedGameName));
            runWineIntent.putExtra("services", getWineServices(selectedGameName));
            runWineIntent.putExtra("virtualDesktop", getWineVirtualDesktop(selectedGameName));
            runWineIntent.putExtra("enableXInput", getEnableXInput(selectedGameName));
            runWineIntent.putExtra("enableDInput", getEnableDInput(selectedGameName));
            runWineIntent.putExtra("cpuAffinity", getCpuAffinity(selectedGameName));

            activity.sendBroadcast(runWineIntent);
            activity.startActivity(runActivityIntent);
        }

        @Override
        public boolean onLongClick(View view) {
            selectedGameName = gameList.get(getAdapterPosition()).name;
            return false;
        }
    }

    public static String selectedGameName = "";

    public static class GameItem {
        public String name;
        public String exePath;
        public String exeArguments;
        public String iconPath;
        public String box64Version;
        public String box64Preset;
        public ArrayList<String> controllersPreset;
        public ArrayList<Boolean> controllersEnableXInput;
        public ArrayList<Boolean> controllersXInputSwapAnalogs;
        public String virtualControllerPreset;
        public boolean virtualControllerEnableXInput;
        public String displayMode;
        public String displayResolution;
        public String vulkanDriver;
        public int vulkanDriverType;
        public String d3dxRenderer;
        public String dxvkVersion;
        public String wineD3DVersion;
        public String vkd3dVersion;
        public boolean wineESync;
        public boolean wineServices;
        public String cpuAffinityCores;
        public boolean wineVirtualDesktop;
        public boolean enableXInput;
        public boolean enableDInput;

        public GameItem(String name, String exePath, String exeArguments, String iconPath) {
            this.name = name;
            this.exePath = exePath;
            this.exeArguments = exeArguments;
            this.iconPath = iconPath;
            this.box64Version = "Global";
            this.box64Preset = "default";
            this.controllersPreset = new ArrayList<>(Arrays.asList("default", "default", "default", "default"));
            this.controllersEnableXInput = new ArrayList<>(Arrays.asList(true, true, true, true));
            this.controllersXInputSwapAnalogs = new ArrayList<>(Arrays.asList(false, false, false, false));
            this.virtualControllerPreset = "default";
            this.virtualControllerEnableXInput = true;
            this.displayMode = "16:9";
            this.displayResolution = "1280x720";
            this.vulkanDriver = "Global";
            this.vulkanDriverType = MESA_DRIVER;
            this.d3dxRenderer = "DXVK";
            this.dxvkVersion = listRatPackagesId("DXVK").get(0);
            this.wineD3DVersion = listRatPackagesId("WineD3D").get(0);
            this.vkd3dVersion = listRatPackagesId("VKD3D").get(0);
            this.wineESync = true;
            this.wineServices = false;
            this.cpuAffinityCores = String.join(",", availableCPUs);
            this.wineVirtualDesktop = false;
            this.enableXInput = true;
            this.enableDInput = true;
        }
    }
}

File: app/src/main/java/com/micewine/emu/adapters/AdapterGame.kt
package com.micewine.emu.adapters

import android.annotation.SuppressLint
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.ColorMatrix
import android.graphics.ColorMatrixColorFilter
import android.graphics.Paint
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.FragmentActivity
import androidx.recyclerview.widget.RecyclerView
import com.micewine.emu.R
import com.micewine.emu.activities.EmulationActivity
import com.micewine.emu.activities.GeneralSettingsActivity.Companion.SELECTED_BOX64
import com.micewine.emu.activities.GeneralSettingsActivity.Companion.SELECTED_VULKAN_DRIVER
import com.micewine.emu.activities.MainActivity.Companion.ACTION_RUN_WINE
import com.micewine.emu.activities.MainActivity.Companion.preferences
import com.micewine.emu.fragments.EditGamePreferencesFragment
import com.micewine.emu.fragments.EditGamePreferencesFragment.Companion.EDIT_GAME_PREFERENCES
import com.micewine.emu.fragments.ShortcutsFragment.Companion.ADRENO_TOOLS_DRIVER
import com.micewine.emu.fragments.ShortcutsFragment.Companion.MESA_DRIVER
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getBox64Preset
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getBox64Version
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getCpuAffinity
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getD3DXRenderer
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getDXVKVersion
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getDisplaySettings
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getSelectedVirtualControllerPreset
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getVKD3DVersion
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getVulkanDriver
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getWineD3DVersion
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getWineESync
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getWineServices
import com.micewine.emu.fragments.ShortcutsFragment.Companion.getWineVirtualDesktop
import java.io.File
import kotlin.math.roundToInt


class AdapterGame(
    private val gameList: MutableList<GameItem>,
    private val size: Float,
    private val activity: FragmentActivity,
) : RecyclerView.Adapter<AdapterGame.ViewHolder>() {
    private var filteredList: MutableList<GameItem> = gameList

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val itemView = LayoutInflater.from(parent.context).inflate(R.layout.adapter_game_item, parent, false)
        itemView.layoutParams.width = (itemView.layoutParams.width * size).roundToInt()
        itemView.layoutParams.height = (itemView.layoutParams.height * size).roundToInt()
        return ViewHolder(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val sList = filteredList[position]
        holder.titleGame.text = sList.name

        val imageFile = File(sList.iconPath)

        if (imageFile.exists() && imageFile.length() > 0) {
            val imageBitmap = BitmapFactory.decodeFile(sList.iconPath)
            if (imageBitmap != null) {
                val newBitmap: Bitmap = if (File(sList.exePath).exists()) {
                    resizeBitmap(
                        imageBitmap, holder.itemView.layoutParams.width - 10, holder.itemView.layoutParams.width - 10
                    )
                } else {
                    toGrayscale(
                        resizeBitmap(
                            imageBitmap, holder.itemView.layoutParams.width - 10, holder.itemView.layoutParams.width - 10
                        )
                    )
                }
                holder.gameImage.setImageBitmap(newBitmap)
            }
        } else if (sList.iconPath == "") {
            holder.gameImage.setImageBitmap(resizeBitmap(
                BitmapFactory.decodeResource(activity.resources, R.drawable.default_icon), holder.itemView.layoutParams.width - 10, holder.itemView.layoutParams.width - 10)
            )
        } else {
            holder.gameImage.setImageBitmap(
                Bitmap.createBitmap(2, 2, Bitmap.Config.ARGB_8888),
            )
        }
    }

    override fun getItemCount(): Int {
        return filteredList.size
    }

    @SuppressLint("NotifyDataSetChanged")
    fun filterList(name: String) {
        filteredList = gameList.filter {
            it.name.contains(name, true)
        }.toMutableList()

        notifyDataSetChanged()
    }

    private fun resizeBitmap(originalBitmap: Bitmap, width: Int, height: Int): Bitmap {
        return Bitmap.createScaledBitmap(originalBitmap, width, height, false)
    }

    private fun toGrayscale(bmpOriginal: Bitmap): Bitmap {
        val bmpGrayscale = Bitmap.createBitmap(bmpOriginal.width, bmpOriginal.height, Bitmap.Config.ARGB_8888)

        Canvas(bmpGrayscale).apply {
            drawBitmap(bmpOriginal, 0f, 0f,
                Paint().apply {
                    colorFilter = ColorMatrixColorFilter(
                        ColorMatrix().apply {
                            setSaturation(0f)
                        }
                    )
                }
            )
            drawRect(0f, 0f, bmpOriginal.width.toFloat(), bmpOriginal.height.toFloat(),
                Paint().apply {
                    color = Color.argb(25, 128, 128, 128)
                }
            )
        }

        return bmpGrayscale
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, View.OnLongClickListener {
        val titleGame: TextView = itemView.findViewById(R.id.title_game_model)
        val gameImage: ImageView = itemView.findViewById(R.id.img_game)

        init {
            itemView.setOnClickListener(this)
            itemView.setOnLongClickListener(this)
        }

        override fun onClick(v: View) {
            val gameModel = gameList[adapterPosition]

            selectedGameName = gameModel.name

            val exeFile = File(gameModel.exePath)
            var exePath = gameModel.exePath
            var exeArguments = gameModel.exeArguments

            if (!exeFile.exists()) {
                if (exeFile.path == activity.getString(R.string.desktop_mode_init)) {
                    exePath = ""
                    exeArguments = ""
                } else {
                    activity.runOnUiThread {
                        Toast.makeText(activity, activity.getString(R.string.executable_file_not_found), Toast.LENGTH_SHORT).show()
                    }

                    EditGamePreferencesFragment(EDIT_GAME_PREFERENCES).show(activity.supportFragmentManager, "")
                    return
                }
            }

            val intent = Intent(activity, EmulationActivity::class.java)

            var driverName = getVulkanDriver(selectedGameName)
            if (driverName == "Global") {
                driverName = preferences?.getString(SELECTED_VULKAN_DRIVER, "").toString()
            }
            var box64Version = getBox64Version(selectedGameName)
            if (box64Version == "Global") {
                box64Version = preferences?.getString(SELECTED_BOX64, "").toString()
            }

            val driverType = if (driverName.startsWith("AdrenoToolsDriver-")) ADRENO_TOOLS_DRIVER else MESA_DRIVER

            val runWineIntent = Intent(ACTION_RUN_WINE).apply {
                putExtra("exePath", exePath)
                putExtra("exeArguments", exeArguments)
                putExtra("driverName", driverName)
                putExtra("driverType", driverType)
                putExtra("box64Version", box64Version)
                putExtra("box64Preset", getBox64Preset(selectedGameName))
                putExtra("displayResolution", getDisplaySettings(selectedGameName)[1])
                putExtra("virtualControllerPreset", getSelectedVirtualControllerPreset(selectedGameName))
                putExtra("d3dxRenderer", getD3DXRenderer(selectedGameName))
                putExtra("wineD3D", getWineD3DVersion(selectedGameName))
                putExtra("dxvk", getDXVKVersion(selectedGameName))
                putExtra("vkd3d", getVKD3DVersion(selectedGameName))
                putExtra("esync", getWineESync(selectedGameName))
                putExtra("services", getWineServices(selectedGameName))
                putExtra("virtualDesktop", getWineVirtualDesktop(selectedGameName))
                putExtra("cpuAffinity", getCpuAffinity(selectedGameName))
            }

            activity.sendBroadcast(runWineIntent)
            activity.startActivity(intent)
        }

        override fun onLongClick(v: View): Boolean {
            val gameModel = gameList[adapterPosition]

            selectedGameName = gameModel.name

            return false
        }
    }

    class GameItem(
        var name: String,
        var exePath: String,
        var exeArguments: String,
        var iconPath: String
    )

    companion object {
        var selectedGameName = ""
    }
}
